Tag templates

alinoeabrassart opened the pull request:
This PR enables you to use container related value inside registrator tags. When you use a {{ <path> }} string inside a container tag, the string is replaced by the value of the <path> key inside docker inspect.

An example is available in compose_examples directory

alinoe-abrassart committed:
Add fetching tags in docker inspect using $ prefix

    Add a function for mapping the tag to the path in docker inspect
    json structure. This add a function that doest the tag mapping
    Corrects some syntax mistakes using go.vim linter
    Should break retrocompatbility only for tags using the dollar
    sign at the beginning

alinoe-abrassart committed:
Add registrator to .gitignore to use it in dev

Commit:
Replaced $ regexp by {{ }} for more flexibility

    Still works for Id and Config.Labels

Commit:
Added compose_examples dir to test registrator and tags

    The following command should allow you to start consul, registrator
    and a test container with and Id tag and a custom tag exposed
    ```
    export GOPATH=<your gopath>
    cd compose_examples
    docker-compose -f tags-example.compose.yml up
    ```
    You need to install docker-compose

mattaitchison commented on bridge/bridge.go:
```diff
@@ -246,6 +246,46 @@ func (b *Bridge) add(containerId string, quiet bool) {
+	// inside {{ }} with the path in the container description
+	// if found, else replace by empty string
+	var tags = make([]string, 0)
+	tag_re := regexp.MustCompile("([^,]+)")
```
Would it be possible to implement this with text.Template? I would rather not introduce regexp if it isn't required.

mattaitchison commented on bridge/bridge.go:
```diff
@@ -246,6 +246,46 @@ func (b *Bridge) add(containerId string, quiet bool) {
+		// for that, we get all braces expression
+		mapped_tag := vp_re.ReplaceAllStringFunc(tag, func(value_path string) string {
+			tag_key := tk_re.FindStringSubmatch(value_path)
+			if tag_key[1] == "Config" {
```
`FindStringSubmatch` could be nil and cause this to panic.

Commit:
Changed to use go-template system

    The tag description gets a bit harder but the code is much
    more readable

alinoeabrassart commented:
Sorry for the absence, the end-of-the-year parties took their tolls. I have made the modification to use the go template system as requested. We can see in the example that tags gets a bit more difficult to write but we gain in code complexity. I'll write some doc on how to use the go template system with the container structure exposed by the go-dockerclient.

Commit:
Added README.md on compose examples

Commit:
Removed unused encoding/json from bridge.go

Commit:
Restore tags

    The previous commits forgot to add the tags to the service
    description, this one adds the missing tag

alinoeabrassart commented:
Finally rebased it against master. @mattaitchison does this version feels better ?

missedone commented on bridge/bridge.go:
```diff
@@ -246,6 +248,21 @@ func (b *Bridge) add(containerId string, quiet bool) {
+	if err != nil {
+		log.Println("executing template:", err)
+	}
+	tags = strings.Split(tag.String(), ",")
```
`tag` could be nil here that would cause panic, since `tagTemplate.Execute` at line 258 could failed with err, in this case, it's better to return the original tags.

Commit:
Add default return in tag template parsing

    This allow to return coherent argument when the templating
    of the tags fails

Commit:
Add error handling when parsing tags templates

    Added few more compose examples to test all tags strings cases

alinoeabrassart commented:
Ok, so before this leaves the first page of pull requests. I've made the intended correction and added examples in compose-example that show the expected behavior when registering to consul. Sorry for the delay.

darkweaver87 commented:
could you please merge it ? this patch is sooooo useful ! :1st_place_medal:

argosback commented:
Hey @alinoeabrassart — the idea from this PR landed in https://github.com/fayrus/registrator, extended to work per container via **SERVICE_TAGS**. You can now use Go templates directly in the **SERVICE_TAGS** label, e.g.
`SERVICE_TAGS=host-{{.Config.Hostname}},web`. It uses the same template functions available in the global -tags flag (from #677). Credit is in the changelog — thanks for pushing this forward.
